applications/FluidDynamicsApplication/custom_elements/alternative_d_vms_dem_coupled.cpp

```diff
@@ -0,0 +1,999 @@
+    array_1d<double,3> convective_velocity = this->FullConvectiveVelocity(rData);
+    BoundedMatrix<double,Dim,Dim> tau_one = ZeroMatrix(Dim,Dim);
+    double fluid_fraction = this->GetAtCoordinate(rData.FluidFraction, rData.N);
+    double tau_two;
+    this->CalculateStabilizationParameters(rData,convective_velocity,tau_one,tau_two);
+
+    const double dt = rData.DeltaTime;
+
+    array_1d<double,3> residual = ZeroVector(3);
+
+    if (rData.UseOSS != 1.0) {
+        this->AlgebraicMomentumResidual(rData,convective_velocity,residual);
+    }
+    else {
+        this->OrthogonalMomentumResidual(rData,convective_velocity,residual);
+    }
+
+    // Note: residual is always of size 3, but stored subscale is of size Dim
+    const auto& rOldSubscaleVelocity = mOldSubscaleVelocity[rData.IntegrationPointIndex];
+    for (unsigned int d = 0; d < Dim; d++) {
+        rVelocitySubscale[d] = tau_one(d,d)*(residual[d] + fluid_fraction * (density/dt)*rOldSubscaleVelocity[d]);
+    }
+}
+
+template< class TElementData >
+void AlternativeDVMSDEMCoupled<TElementData>::SubscalePressure(
+    const TElementData& rData,
+    double &rPressureSubscale) const
+{
+    array_1d<double,3> convective_velocity = this->FullConvectiveVelocity(rData);
+
+    BoundedMatrix<double,Dim,Dim> tau_one = ZeroMatrix(Dim,Dim);
+    double tau_two;
+    this->CalculateStabilizationParameters(rData,convective_velocity,tau_one,tau_two);
+
+    double residual = 0.0;
+
+    if (rData.UseOSS != 1.0)
+        this->AlgebraicMassResidual(rData,residual);
+    else
+        this->OrthogonalMassResidual(rData,residual);
+
+    rPressureSubscale = tau_two*residual;
+}
+
+template< class TElementData >
+array_1d<double,3> AlternativeDVMSDEMCoupled<TElementData>::FullConvectiveVelocity(
+    const TElementData& rData) const
+{
+    array_1d<double,3> convective_velocity = this->GetAtCoordinate(rData.Velocity,rData.N) - this->GetAtCoordinate(rData.MeshVelocity,rData.N);
+    // Adding subscale term componentwise because return type is of size 3, but subscale is of size Dim
+    const array_1d<double,Dim>& r_predicted_subscale = mPredictedSubscaleVelocity[rData.IntegrationPointIndex];
+
+    for (unsigned int d = 0; d < Dim; d++) {
+        convective_velocity[d] += r_predicted_subscale[d];
+    }
+
+    return convective_velocity;
+}
+
+template< class TElementData >
+void AlternativeDVMSDEMCoupled<TElementData>::UpdateSubscaleVelocity(
+    const TElementData& rData)
+{
+    const double density = this->GetAtCoordinate(rData.Density,rData.N);
+    double fluid_fraction = this->GetAtCoordinate(rData.FluidFraction,rData.N);
+    array_1d<double,3> resolved_convection_velocity = this->FullConvectiveVelocity(rData);
+    const double dt = rData.DeltaTime;
+    array_1d<double,3> predicted_subscale_velocity = ZeroVector(3);
+
+    const array_1d<double,Dim>& old_subscale_velocity = mOldSubscaleVelocity[rData.IntegrationPointIndex];
```

Comment:
```suggestion
    const array_1d<double,Dim>& r_old_subscale_velocity = mOldSubscaleVelocity[rData.IntegrationPointIndex];
```
(minor style guide comment)